src/vendored/CowWrapper.sol

```diff
@@ -0,0 +1,178 @@
+        uint256 buyAmount;
+        uint32 validTo;
+        bytes32 appData;
+        uint256 feeAmount;
+        uint256 flags;
+        uint256 executedAmount;
+        bytes signature;
+    }
+
+    /// @notice Interaction data structure for pre/intra/post-settlement hooks
+    struct CowInteractionData {
+        address target;
+        uint256 value;
+        bytes callData;
+    }
+
+    /// @notice Settles a batch of trades atomically
+    /// @param tokens Array of token addresses involved in the settlement
+    /// @param clearingPrices Array of clearing prices for each token
+    /// @param trades Array of trades to execute
+    /// @param interactions Array of three interaction arrays (pre, intra, post-settlement)
+    function settle(
+        address[] calldata tokens,
+        uint256[] calldata clearingPrices,
+        CowTradeData[] calldata trades,
+        CowInteractionData[][3] calldata interactions
+    ) external;
+
+    function vaultRelayer() external returns (address);
+}
+
+/// @title CoW Protocol Wrapper Interface
+/// @notice Interface for wrapper contracts that add custom logic around CoW settlements
+/// @dev Wrappers can be chained together to compose multiple settlement operations
+interface ICowWrapper {
+    /// @notice Initiates a wrapped settlement call
+    /// @dev This is the entry point for wrapped settlements. The wrapper will execute custom logic
+    ///      before calling the next wrapper or settlement contract in the chain.
+    /// @param settleData ABI-encoded call to CowSettlement.settle()
+    /// @param wrapperData Encoded chain of wrapper-specific data followed by addresses of next wrappers/settlement
+    function wrappedSettle(
+        bytes calldata settleData,
+        bytes calldata wrapperData
+    ) external;
+
+    /// @notice Parses and validates wrapper-specific data
+    /// @dev Used by CowWrapperHelpers to validate wrapper data before execution.
+    ///      Implementations should consume their portion of wrapperData and return the rest.
+    /// @param wrapperData The wrapper-specific data to parse
+    /// @return remainingWrapperData Any wrapper data that was not consumed by this wrapper
+    function parseWrapperData(bytes calldata wrapperData) external view returns (bytes calldata remainingWrapperData);
+}
+
+/// @title CoW Protocol Wrapper Base Contract
+/// @notice Abstract base contract for creating wrapper contracts around CoW Protocol settlements
+/// @dev A wrapper enables custom pre/post-settlement and context-setting logic and can be chained with other wrappers.
+///      Wrappers must:
+///      - Be approved by the CowAuthentication contract
+///      - Verify the caller is an authenticated solver
+///      - Eventually call settle() on the approved CowSettlement contract
+///      - Implement _wrap() for custom logic
+///      - Implement parseWrapperData() for validation of implementation-specific wrapperData
+abstract contract CowWrapper {
```

Comment:
For PoC is good, but I still think we should have both the abstract contract and interface. We don't want to limit the way people make wrappers and we prefer working wirth interfaces to apply SOLID principles in our contract designs